Doc/library/sqlite3.rst

```diff
@@ -2358,6 +2340,115 @@ can be found in the `SQLite URI documentation`_.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` represent fetched rows as :class:`tuples <tuple>`.
+If a :class:`!tuple` does not suit your needs,
+use the :class:`sqlite3.Row` class or a custom :attr:`~Connection.row_factory`.
+
+:class:`!Row` provides indexed and case-insensitive named access to columns,
+with low memory overhead and minimal performance impact.
+In order to use :class:`!Row` as a row factory,
+simply assign it to the :attr:`Connection.row_factory` attribute:
+
+.. doctest::
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = sqlite3.Row
+
+Query results are now returned as :class:`!Row` instances:
+
+.. doctest::
+
+   >>> res = con.execute("SELECT 'Earth' AS name, 6378 AS radius")
+   >>> row = res.fetchone()
+   >>> row.keys()
+   ['name', 'radius']
+   >>> row[0]  # Access by index.
+   'Earth'
+   >>> row["name"]    # Access by name.
+   'Earth'
+   >>> row["RADIUS"]  # Column names are case-insensitive.
+   6378
+
+It is also possible to assign row factories to cursors using
+:attr:`Cursor.row_factory`:
+
+.. doctest::
+
+   >>> cur = con.cursor()
+   >>> cur.row_factory == con.row_factory
+   True
+   >>> cur.row_factory = None  # Override cursor row factory.
+
+   # The cursor and the connection row factories now differ.
+   >>> cur.row_factory == con.row_factory
+   False
+   >>> cur.execute("SELECT 'Hello'").fetchone()
+   ('Hello',)
+
+   # The connection still uses sqlite3.Row, as set in the previous example.
+   >>> con.execute("SELECT 'Hello'").fetchone()
+   <sqlite3.Row object at ...>
+
+If more flexibility is needed, implement a custom row factory.
+Here's an example of one returning a :class:`dict`:
+
+.. doctest::
+
+   >>> def dict_factory(cursor, row):
+   ...     col_names = [col[0] for col in cursor.description]
+   ...     return {key: value for key, value in zip(col_names, row)}
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = dict_factory
+   >>> for row in con.execute("SELECT 1 AS a, 2 AS b"):
+   ...     print(row)
+   {'a': 1, 'b': 2}
+
+Here's an example of a factory that return a :class:`~collections.namedtuple`.
+
+.. testcode::
+
+   from collections import namedtuple
+
+   def _get_fields(cursor):
+        return (col[0] for col in cursor.description)
+
+   def _make_cls(fields):
+       return namedtuple("Row", fields)
+
+   def namedtuple_factory(cursor, row):
+       cls = _make_cls(_get_fields(cursor))
+       return cls._make(row)
+
+:func:`!namedtuple_factory` can be used as follows:
+
+.. doctest::
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = namedtuple_factory
+   >>> cur = con.execute("SELECT 1 AS a, 2 AS b")
+   >>> row = cur.fetchone()
+   >>> row
+   Row(a=1, b=2)
+   >>> row[0]  # Indexed access.
+   1
+   >>> row.b   # Attribute access.
+   2
+
+As an optimisation, decorate :func:`!_make_cls` with
+:func:`!functools.lru_cache` to avoid creating multiple named tuples
```

Comment:
```suggestion
:func:`!functools.lru_cache` to avoid creating multiple :class:`!namedtuple`\s
```

Clarify that this is the `namedtuple` class, not just a `tuple` that has a name.

Reply:
Is this the first time named tuples are mentioned?  Is it a real worry that people will not know about collections.namedtuple but imagine a concept of tuple that has a name?

If namedtuple was already linked before, then the previous markup was fine IMO.  If it’s the first time, we have ``` :term:`named tuple` ``` to link to glossary.

Reply:
> Is this the first time named tuples are mentioned?

No, otherwise I wouldn't have added `!` :)

> Is it a real worry that people will not know about collections.namedtuple but imagine a concept of tuple that has a name?

Probably not—as you point out, that wasn't a very good reason, heh.

Your reference to the [glossary entry](https://docs.python.org/3.12/glossary.html#term-named-tuple) (that I forgot existed) helps clarify the distinction here between "named tuples" and the `collections.namedtuple` class factory. The former are defined there as any objects that inherent from tuple and offer attribute access to named fields, which may be created manually, via `collections.namedtuple`, or other means.

Here we're specifically using the latter, and the cost in question is not that of creating multiple _instances_ of named tuple objects, but rather calling `collections.namedtuple` multiple times to construct the class for the named tuples. Also, the defining feature of named tuples mentioned there (access by attribute) is not actually demonstrated in the examples.

With that in mind, this can be revised to reflect this, as well as further clarify the effect.

```suggestion
:func:`!functools.lru_cache` to avoid calling :func:`namedtuple` many times
with the same row schema.
```

Reply:
It's mentioned at line 2412.  In general I would be a bit more lenient with the use of `!` -- unless there is another link in the immediate vicinity, I would avoid adding it, and it's probably better to have an extra link than not having one.

Maybe in the rendered HTML it's easier to spot, but I haven't seen a link to `lru_cache`.  An advantage of using the link is also that you can use `` :func:`~functools.lru_cache` `` and hide the module, since the link will already bring the user to the right page.

Reply:
Thanks. I ensured `collections.namedtuple` is linked, since the last place it is mentioned is above the example block. IIRC there was a link to `functools.lru_cache` above the example, but that was removed when that optimization was removed from the given example in the most recent revision, and we missed removing the `!` here, thanks. I've edited my suggestion to fix that.